ugs-core/src/com/willwinder/universalgcodesender/MarlinController.java

```diff
@@ -0,0 +1,409 @@
+		return null;
+	}
+
+	@Override
+	public ControllerStatus getControllerStatus() {
+		return controllerStatus;
+	}
+
+	@Override
+	protected Boolean isIdleEvent() {
+		// TODO: what should it do - GRBL does some real time check here
+		return true;
+	}
+
+	@Override
+	protected void closeCommBeforeEvent() {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	protected void closeCommAfterEvent() {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	protected void cancelSendBeforeEvent() throws Exception {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	protected void cancelSendAfterEvent() throws Exception {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	protected void pauseStreamingEvent() throws Exception {
+		logger.info("trying to send M0");
+		GcodeCommand command = createCommand("M0");
+		sendCommandImmediately(command);
+	}
+
+	@Override
+	protected void resumeStreamingEvent() throws Exception {
+		logger.info("trying to send M108");
+		GcodeCommand command = createCommand("M108");
+		// sendCommandImmediately(command);
+		this.marlinComm.sendRealtimeCommand(command);
+
+		synchronized (this) {
+			// need to resume otherwise the cmd will never go
+			isResuming = true;
+			comm.resumeSend();
+		}
+	}
+
+	@Override
+	protected void isReadyToSendCommandsEvent() throws Exception {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	protected void isReadyToStreamCommandsEvent() throws Exception {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	protected void rawResponseHandler(String response) {
+		String processed = response;
+		try {
+			boolean verbose = false;
+
+			if (MarlinUtils.isOkResponse(response)) {
+				this.commandComplete(processed);
+				logger.info("active count after rx: " + marlinComm.activeCommandListSize());
+				updateControllerState("Idle", ControllerState.IDLE);
+				marlinComm.setMarlinBusy(false);
+				isResuming = false;
+			} else if (MarlinUtils.isPausedResponse(response)) {
+				updateControllerState("Paused", ControllerState.HOLD);
+				marlinComm.setMarlinBusy(false);
+				synchronized (this) {
+					if (!isResuming) {
+						comm.pauseSend();
+					}
+				}
+				// dont stop sending, need to be able to resume with M108
+			} else if (MarlinUtils.isBusyResponse(response)) {
+				// do what?
+				// marlinComm.setMarlinBusy(true);
+				updateControllerState("Run", ControllerState.RUN);
+			} else if (MarlinUtils.isMarlinStatusString(response)) {
+				// Only 1 poll is sent at a time so don't decrement, reset to zero.
+				this.outstandingPolls = 0;
+
+				// Status string goes to verbose console
+				verbose = true;
+
+				// this.handleStatusString(response);
+
+				controllerStatus = MarlinUtils.getStatusFromStatusString(
+						controllerStatus, response, capabilities, getFirmwareSettings().getReportingUnits());
+
+				dispatchStatusString(controllerStatus);
+
+				this.checkStreamFinished();
+			} else if (response.contains("Free Memory:")) {
```

Comment:
Is this the best way to identify that it is connected to a Marlin controller?

When I experimented with this I used the "start" message:
https://github.com/winder/Universal-G-Code-Sender/compare/master...breiler:feature/marlin-firmware-2#diff-4efdcf4b045d1fc9c9610b760d4a7589fa3a7ab94273ea524d6b8821e64c4b65R109